### ISSUE TITLE
VEDA Attribute variants possibly not correctly characterized in veda-attr-defaults?

### ISSUE THREAD

**Issue by Antti-L** [etsap-TIMES/xl2times#225]
The `CEFF`, `CEFFICIENCY`, `CEFF-I` and `CEFF-O` attributes are VEDA variants of the TIMES `ACT_EFF` attribute.  However, they are not aliases, because `CEFF`, `CEFFICIENCY`, `CEFF-I` and `CEFF-O` accept only commodities.  `Other_Indexes` should thus be **ignored** for these attributes (it is ignored under VEDA). On the other hand, `ACT_EFF` supports any `CG`, and `Other_Indexes` does take the highest priority, and so if e.g. both `Other_Indexes` and `CSet_CN` are filled in, the `CG` index is taken from `Other_Indexes`.

Similarly, the `INPUT` and `OUTPUT` attributes are VEDA variants of the TIMES `VDA_FLOP` attribute.  Again, they are not aliases, because `INPUT` and `OUTPUT` accept only commodities.  `Other_Indexes` should thus be **ignored** also for these attributes (it is ignored under VEDA).  And here again, `VDA_FLOP` supports any `CG`,  and `Other_Indexes` takes the highest priority for the `CG` index.

Notwithstanding, I can see that in the `veda-attr-defaults` file, all these attributes seem to be characterized as if they would be supporting the `Other_Indexes` field/column.  I am not sure I am interpreting that correctly, but it looks suspicious to me.  If those defaults impliy that a commodity index might be taken from `Other_Indexes` for `CEFF`, `CEFFICIENCY`, `CEFF-I`, `CEFF-O`, `INPUT` or `OUTPUT`, that would lead to incorrect model input data.  @olejandro can you confirm whether this might be the case or not?

**Comment by olejandro** [comment]
@Antti-L thanks for a great question! I won't go into detail on the `ACT_EFF` variants yet, because they are pending revision.

Let me for now just explain what the data structure in `veda-attr-defaults` means. Meanwhile we naturally keep this issue open until of the concerns are addressed.

Below is an entry for `INPUT`. `times-attribute` field identifies it as a variant of `VDA_FLOP`, while the `defaults` section indicates how to fill the `other_indexes` entry which is currently mapped to `cg` index of `VDA_FLOP` in `times-info.json`. As far as I understand your concern is related to a possibility that `other_indexes` may be not empty and so the `defaults` field won't be applied. Even though we have not seen this in our benchmarks, this is a vaild issue. At the same, the use of `other_indexes` in the internal data structures is temporary and will be soon changed to avoid issue, ensuring that the index is always filled correctly.

```
"INPUT": {
    "defaults": {
      "other_indexes": [
        "commodity",
        "commodity-in",
        "commodity-in-aux"
      ],
      "ts-level": "ANNUAL"
    },
    "times-attribute": "VDA_FLOP"
  }
```

**Comment by olejandro** [comment]
After the change the entries will look similar to below:

```
"INPUT": {
    "defaults": {
      "new_internal_name": [
        "commodity",
        "commodity-in",
        "commodity-in-aux"
      ],
      "ts-level": "ANNUAL"
    },
    "times-attribute": "VDA_FLOP"
  }
```

```
"VDA_FLOP": {
    "defaults": {
      "new_internal_name": [
        "other_indexes",
        "commodity",
        "commodity-in",
        "commodity-out",
        "commodity-in-aux",
        "commodity-out-aux"
      ],
      "ts-level": "ANNUAL"
    }
  }
```

**Comment by Antti-L** [comment]
> As far as I understand your concern is related to a possibility that other_indexes may be not empty and so the defaults field won't be applied. 

Exactly.  As far as I can tell, that situation could be reasonably common in some models (in my models for sure).  Thanks for confirming that it is currently an outstanding issue, and great if you can already say this issue will be resolved.

**Comment by Antti-L** [comment]
@olejandro Small follow-up:
It is now over **9 months** since the original discussion, and so I did some tests now to see if the issues are still outstanding.  
I can confirm that the issues are indeed still present:  For example, `INPUT` and `CEFF` are getting incorrectly defined under `xl2times`.  I tested the same specs with VEDA2 and they worked in the way expected (i.e. `Other_indexes` is not taken into account for `INPUT` and `CEFF`, because they are "flow" attributes and thus expect a **commodity** only).

On the other hand, under `xl2times` `ACT_EFF`, `ACT_FLO` and `PRC_ACTFLO` are getting nothing defined if they are specified with `CSet_CN`, which is inconsistent behaviour.  These attributes accept **any CG**, and thus a commodity in `CSet_CN` must be accepted whenever **Other_indexes** has **not** been specified.  Again here, I verified that VEDA2 behaves in the expected way.  However, I can see that `VDA_FLOP` does already work correctly in this respect (it does accept `CSet_CN`), and so I guess this inconsistency would be easy to fix:  Just define the defaults for `ACT_EFF`, `ACT_FLO` and `PRC_ACTFLO` pretty much in the same way as for `VDA_FLOP`: 
 ```
   "defaults": {
      "other_indexes": [
        "commodity",
        "commodity-in",
        "commodity-out",
        "commodity-in-aux",
        "commodity-out-aux"
      ],
      "ts-level": "ANNUAL"
    }
```
Except that for `ACT_EFF` and `PRC_ACTFLO` the auxiliary commodities should apparently not be included.

**Comment by olejandro** [comment]
Thanks @Antti-L!
I've made some adjustments here: https://github.com/etsap-TIMES/xl2times/pull/273/files. Does it look right?
I guess, I should remove aux flows from all `ACT_EFF` aliases/variants in the file (i.e. `CEFF`, `CEFF-I`, `CEFF-O`, `CEFFICIENCY`, `EFF`, and `EFFICIENCY`), as well?
I see there is also `VDA_EFF` included in the file. Is it an alias or a variant?
What about `ACT2FLO` and `ACTFLO`? Are they just aliases of `PRC_ACTFLO` or variants?
Regarding the `INPUT` and `CEFF` issue, I should be able to get to it soon.

**Comment by Antti-L** [comment]
> Does it look right?

Yes, I guess so. But see also below.

> I should remove aux flows from all ACT_EFF aliases/variants in the file (i.e. CEFF, CEFF-I, CEFF-O, CEFFICIENCY, EFF, and EFFICIENCY), as well?

Well, that's what I tended to suggest, because VEDA defines a **zero ACT_EFF for any auxiliary flows**, in order to **remove** these flows from the efficiency group (because of the nature of auxiliary flows).  However, see also https://github.com/etsap-TIMES/xl2times/pull/155#issuecomment-1861799117 . VEDA may indeed still be supporting ACT_EFF/CEFF for commodities specified under TOP-IN-A TOP-OUT-A as well, and so it might be a matter of some further consideration (some models might perhaps even exploit it?).  But as said, in my understanding auxiliary flows are basically thought **not to be members** of ACT_EFF efficiency groups, which is why the zero ACT_EFFs are being generated for them.

But concerning EFF and EFFICIENCY, they are **variants** that do not accept any commodity or CG, and I think that should be the case also under xl2times.  When converting them to `ACT_EFF`, `ACT` should just be unconditionally filled-in.

* VDA_EFF seems to be an alias of EFF, i.e alias of the variant that has no commodity or CG index.
* ACT2FLO and ACTFLO seem to be just aliases of PRC_ACTFLO

**Comment by olejandro** [comment]
Great, thanks @Antti-L!
I've removed the aux flow for now. We can always add them back later if necessary.
Regarding `EFF` and the likes, sorry, my bad, it is already set up to behave that way once we add the new dimension to the internal data structure. I've added same defaults to `VDA_EFF`
